
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>The Data &#8212; Ecoevolity 0.1.0 documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="The Configuration File" href="yaml-config.html" />
    <link rel="prev" title="Background" href="background.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../index.html">
    <img class="logo" src="../_static/site-logo.svg" alt="Logo"/>
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=phyletica&repo=ecoevolity&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Ecoevolity</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="background.html">Background</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">The Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="yaml-config.html">The Configuration File</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../phycoeval/index.html">Phycoeval</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshops/index.html">Workshops</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Change History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citation.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../acknowledgments.html">Acknowledgments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zbib/references.html">Literature Cited</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html"></a><ul>
      <li>Previous: <a href="background.html" title="previous chapter">Background</a></li>
      <li>Next: <a href="yaml-config.html" title="next chapter">The Configuration File</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <p><a class="reference external" href="index.html"><img alt="Ecoevolity" src="../_images/ecoevolity-logo.svg" width="95%" /></a></p>
<div class="section" id="the-data">
<span id="data"></span><h1>The Data<a class="headerlink" href="#the-data" title="Permalink to this headline">¶</a></h1>
<p>To use <a class="reference external" href="http://phyletica.org/ecoevolity">ecoevolity</a>, we need to have sampled genetic data from populations we are
interested in comparing.
Ecoevolity assumes your genetic characters are:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#orthologous"><span class="std std-ref">orthologous</span></a></p></li>
<li><p><a class="reference internal" href="#biallelic"><span class="std std-ref">biallelic</span></a></p></li>
<li><p><a class="reference internal" href="#unlinked"><span class="std std-ref">effectively unlinked</span></a></p></li>
</ul>
<div class="section" id="orthology">
<span id="orthologous"></span><h2>Orthology<a class="headerlink" href="#orthology" title="Permalink to this headline">¶</a></h2>
<p><strong>Orthologous</strong> means that you are looking at the “same” character across your
sampled genomes; the character was inherited from a common ancestor without any
duplications of the locus.
<strong>Orthology is only assumed within each comparison</strong>.
For example, if you are comparing the divergence times between two pairs of
species, the characters do not have to be orthologous across both pairs of
species.
You can take advantage of this if you are assembling your data from raw reads.
For example, if you have “RADseq” data, you can assemble your loci for each
comparison separately to increase the amount of data you end up with.</p>
</div>
<div class="section" id="biallelic">
<span id="id1"></span><h2>Biallelic<a class="headerlink" href="#biallelic" title="Permalink to this headline">¶</a></h2>
<p><strong>Biallelic</strong> means that each character has two possible states.
But, most genomic data are composed of nucleotides, which have four states.
This can be accommodated by
<a class="reference internal" href="#codingnucs"><span class="std std-ref">coding the nucleotides as biallelic</span></a>.</p>
<div class="section" id="coding-nucleotide-characters">
<span id="codingnucs"></span><h3>Coding nucleotide characters<a class="headerlink" href="#coding-nucleotide-characters" title="Permalink to this headline">¶</a></h3>
<p>Ecoevolity will read your nucleotide characters, and recode them as biallelic.
If ecoevolity finds a polyallelic character with more than two states across your
samples, it will report an error and stop running.
However, you can tell ecoevolity to recode these polyallelic characters as either
being the first state (0) or a different state (1).
Your other option is to exclude these sites.
For the data we have analyzed so far, there have been no discernible
differences in the results between recoding or excluding characters with more
than 2 states.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are analyzing nucleotide data as biallelic characters, we strongly
recommend that you do not try and estimate the frequencies of the two
states.
Instead, <a class="reference internal" href="yaml-config.html#freq-1"><span class="std std-ref">fix the frequencies to be equal</span></a>.
This is because there are many ways to recode the 4 nucleotide states to two
states.
Thus, if you try to estimate frequencies of the two states, your results can be
sensitive to how you decided to code your nucleotides as binary.</p>
</div>
</div>
</div>
<div class="section" id="unlinked">
<span id="id2"></span><h2>Unlinked<a class="headerlink" href="#unlinked" title="Permalink to this headline">¶</a></h2>
<p><strong>Unlinked</strong> means that each character is assumed to have evolved along a gene
tree that is independent from the others (conditional on the population
history).
In other words, it assumes that the characters are far enough apart from one
another in the genome that they segregate independently during meiosis.</p>
<p>Many genomic data sets consist of many loci, each of which is a stretch of
linked nucleotides.
Examples include “RADseq” or sequence-capture data.
<a class="reference internal" href="#linkage"><span class="std std-ref">What should we do with such data</span></a>?</p>
<div class="section" id="what-to-do-with-linked-characters">
<span id="linkage"></span><h3>What to do with linked characters?<a class="headerlink" href="#what-to-do-with-linked-characters" title="Permalink to this headline">¶</a></h3>
<p>What if you have loci comprising sequences of linked nucleotides?
Based on simulations of loci of 100, 500, and 1000 linked characters
<a class="bibtex reference internal" href="../zbib/references.html#oaks2018ecoevolity" id="id3">[14]</a>, we strongly recommend that you analyze all of your
characters (including the constant ones) and violate the assumption of unlinked
characters.
Ecoevolity performs much better when you use all of the sites (especially the
constant ones) compared to reducing the data to only one variable character per
locus.
So, your best bet is to put all of your loci for a comparison together into a
nexus-formatted character matrix.</p>
</div>
</div>
<div class="section" id="formatting-your-data">
<span id="nexusfile"></span><h2>Formatting your data<a class="headerlink" href="#formatting-your-data" title="Permalink to this headline">¶</a></h2>
<p>Currently, ecoevolity accepts two input formats for genetic characters:
Nexus and <a class="reference external" href="http://yaml.org/">YAML</a>.
The character data for each of your comparisons will go into a separate
file in one of these formats, which are described below.</p>
<div class="section" id="nexus-format">
<h3>Nexus format<a class="headerlink" href="#nexus-format" title="Permalink to this headline">¶</a></h3>
<div class="section" id="standard-haploid-data">
<h4>Standard haploid data<a class="headerlink" href="#standard-haploid-data" title="Permalink to this headline">¶</a></h4>
<p>You can represent your data in a “standard” 0/1 format.
Here’s example of a pair of populations from which we’ve sampled 4 genomes (2
diploid individuals) from two different populations (indicated by the last part
of the taxon label):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN TAXA;
    DIMENSIONS NTAX=8;
    TAXLABELS
        RMB-5953-a-BabuyanClaro
        RMB-5953-b-BabuyanClaro
        RMB-5954-a-BabuyanClaro
        RMB-5954-b-BabuyanClaro
        RMB-6052-a-Calayan
        RMB-6052-b-Calayan
        RMB-6054-a-Calayan
        RMB-6054-b-Calayan
    ;
END;

BEGIN CHARACTERS;
    DIMENSIONS NCHAR=273658;
    FORMAT DATATYPE=STANDARD SYMBOLS=&quot;01&quot; MISSING=? GAP=-;
    MATRIX
        RMB-5953-a-BabuyanClaro     001010...
        RMB-5953-b-BabuyanClaro     001010...
        RMB-5954-a-BabuyanClaro     101010...
        RMB-5954-b-BabuyanClaro     001011...
        RMB-6052-a-Calayan          101110...
        RMB-6052-b-Calayan          101110...
        RMB-6054-a-Calayan          001011...
        RMB-6054-b-Calayan          101010...
    ;
END;
</pre></div>
</div>
<p>Note, we don’t need separate TAXA and CHARACTER blocks like above.
Instead, we can specify a DATA block:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=8 NCHAR=273658;
    FORMAT DATATYPE=STANDARD SYMBOLS=&quot;01&quot; MISSING=? GAP=-;
    MATRIX
        RMB-5953-a-BabuyanClaro     001010...
        RMB-5953-b-BabuyanClaro     001010...
        RMB-5954-a-BabuyanClaro     101010...
        RMB-5954-b-BabuyanClaro     001011...
        RMB-6052-a-Calayan          101110...
        RMB-6052-b-Calayan          101110...
        RMB-6054-a-Calayan          001011...
        RMB-6054-b-Calayan          101010...
    ;
END;
</pre></div>
</div>
<p>Both examples above would be equivalent to ecoevolity, but the
<a class="reference external" href="http://ncl.sourceforge.net/">Nexus Class Library</a>
used by ecoevolity will report a message about an implicit TAXA block if you use the
latter format.
Either way, in your <a class="reference internal" href="yaml-config.html#configfile"><span class="std std-ref">ecoevolity config file</span></a>,
you need to tell ecoevolity that the
<a class="reference internal" href="yaml-config.html#genotypesarediploid"><span class="std std-ref">states, or genotypes, are haploid</span></a>
by declaring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">false</span>
</pre></div>
</div>
</div>
<div class="section" id="standard-diploid-data">
<h4>Standard diploid data<a class="headerlink" href="#standard-diploid-data" title="Permalink to this headline">¶</a></h4>
<p>Above, each cell in our matrix represented which state was present
for the character in a particular haploid genome.
We can also represent the same data where each cell represents
the genotype of a diploid individual:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=4 NCHAR=273658;
    FORMAT DATATYPE=STANDARD SYMBOLS=&quot;012&quot; MISSING=? GAP=-;
    MATRIX
        RMB-5953-BabuyanClaro     002020...
        RMB-5954-BabuyanClaro     102021...
        RMB-6052-Calayan          202220...
        RMB-6054-Calayan          101021...
    ;
END;
</pre></div>
</div>
<p>Now, “0” represents that the individual has two copies with the 0 state, “2”
represents two copies of the 1 state, and “1” represents a heterozygote.
Again, in your <a class="reference internal" href="yaml-config.html#configfile"><span class="std std-ref">ecoevolity config file</span></a>,
you need to tell ecoevolity that the
<a class="reference internal" href="yaml-config.html#genotypesarediploid"><span class="std std-ref">states, or genotypes, are diploid</span></a>
by declaring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">true</span>
</pre></div>
</div>
</div>
<div class="section" id="nucleotide-data">
<h4>Nucleotide data<a class="headerlink" href="#nucleotide-data" title="Permalink to this headline">¶</a></h4>
<p>If you have nucleotide data, the easiest thing is provide the nucleotide
characters to ecoevolity as is, and let it recode them as biallelic.
Here’s an example where we are providing nucleotides as haploid (each cell is a
haploid genotype):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=8 NCHAR=273658;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        RMB-5953-a-BabuyanClaro     ACGTAG...
        RMB-5953-b-BabuyanClaro     ACGTAG...
        RMB-5954-a-BabuyanClaro     GCGTAG...
        RMB-5954-b-BabuyanClaro     ACGTAA...
        RMB-6052-a-Calayan          GCGCAG...
        RMB-6052-b-Calayan          GCGCAG...
        RMB-6054-a-Calayan          ACGTAA...
        RMB-6054-b-Calayan          GCGTAG...
    ;
END;
</pre></div>
</div>
<p>This is sometimes referred to as “phased” data.
Again, if we are providing a matrix where each cell represents a haploid genotype,
we need to tell ecoevolity this is so via the
<a class="reference internal" href="yaml-config.html#configfile"><span class="std std-ref">config file</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">false</span>
</pre></div>
</div>
<p>We can also represent the same data as “unphased”, where each cell represents a
diploid genotype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=4 NCHAR=273658;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        RMB-5953-BabuyanClaro     ACGTAG...
        RMB-5954-BabuyanClaro     RCGTAR...
        RMB-6052-Calayan          GCGCAG...
        RMB-6054-Calayan          RCGTAR...
    ;
END;
</pre></div>
</div>
<p>We need to indicate this in the
<a class="reference internal" href="yaml-config.html#configfile"><span class="std std-ref">config file</span></a>::
accordingly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">true</span>
</pre></div>
</div>
</div>
<div class="section" id="population-labels">
<h4>Population labels<a class="headerlink" href="#population-labels" title="Permalink to this headline">¶</a></h4>
<p>In our nexus character matrix, we need to indicate which population each
row corresponds to.
We can do this with either using a prefix or suffix in the row (or taxon)
labels.
For example, in this example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=4 NCHAR=273658;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        RMB-5953-BabuyanClaro     ACGTAG...
        RMB-5954-BabuyanClaro     RCGTAR...
        RMB-6052-Calayan          GCGCAG...
        RMB-6054-Calayan          RCGTAR...
    ;
END;
</pre></div>
</div>
<p>we are using the suffixes to indicate that the first two samples came
from a population we are calling <code class="docutils literal notranslate"><span class="pre">BabuyanClaro</span></code>, and
the last two samples came from a population we are calling
<code class="docutils literal notranslate"><span class="pre">Calayan</span></code>.
In our <a class="reference internal" href="yaml-config.html#configfile"><span class="std std-ref">|eco| config file</span></a>
we have to indicate this with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">population_name_delimiter</span><span class="p">:</span> <span class="s2">&quot;-&quot;</span>
<span class="n">population_name_is_prefix</span><span class="p">:</span> <span class="n">false</span>
</pre></div>
</div>
<p>This tells ecoevolity to look for the last bit of each row label that is
separated by a “-” to figure out the population label.</p>
<p>Every nexus file must one or two population labels.
If ecoevolity finds two population labels, it will model
the comparison as two diverged populations and try to
estimate the time that they diverged:</p>
<a class="reference internal image-reference" href="../_images/div-model-singleton.svg"><img alt="divergence comparison cartoon" class="align-center" src="../_images/div-model-singleton.svg" width="50%" /></a>
<p>If ecoevolity finds one population label, it will
model the comparison as a population that underwent
a population-size change and try to estimate the time
that the change occurred:</p>
<a class="reference internal image-reference" href="../_images/demog-model-singleton.svg"><img alt="demog comparison cartoon" class="align-center" src="../_images/demog-model-singleton.svg" width="50%" /></a>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you like to use underscores as a population label
delimiter, just watch out for a
<a class="reference internal" href="yaml-config.html#underscoregotcha"><span class="std std-ref">gotcha related to how the nexus format treats underscores</span></a></p>
</div>
</div>
</div>
<div class="section" id="yaml-format">
<h3>YAML format<a class="headerlink" href="#yaml-format" title="Permalink to this headline">¶</a></h3>
<p>The YAML-formatted data format is a lot more efficient (much smaller file
sizes).
Instead of a full alignment,
it only contains a list of allele count patterns, followed by a list of the
weight of each pattern (i.e., how many times the allele pattern occurs in the
alignment).
For a very small example let’s convert the following nexus alignment
and convert it to the <a class="reference external" href="http://yaml.org/">YAML</a> format that ecoevolity accepts as input:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=8 NCHAR=6;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  ACGTAG
        lizard-953-b-speciesA  ACGTAG
        lizard-954-a-speciesA  GCGTAG
        lizard-954-b-speciesA  ACGTAA
        lizard-152-a-speciesB  GCGCAG
        lizard-152-b-speciesB  GCGCAG
        lizard-154-a-speciesB  ACGTAA
        lizard-154-b-speciesB  GCGTAG
    ;
END;
</pre></div>
</div>
<p>To convert these data to <a class="reference external" href="http://yaml.org/">YAML</a> format, we will assume the first nucleotide
(from the top) in each column is state “0”, and the second nucleotide (if any)
is state “1”.
Doing so gives us the following allele-count patterns in <a class="reference external" href="http://yaml.org/">YAML</a> format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">---</span>
<span class="n">markers_are_dominant</span><span class="p">:</span> <span class="n">false</span>
<span class="n">population_labels</span><span class="p">:</span>
    <span class="o">-</span> <span class="n">speciesA</span>
    <span class="o">-</span> <span class="n">speciesB</span>
<span class="n">allele_count_patterns</span><span class="p">:</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
<span class="n">pattern_weights</span><span class="p">:</span>
    <span class="o">-</span> <span class="mi">1</span>
    <span class="o">-</span> <span class="mi">3</span>
    <span class="o">-</span> <span class="mi">1</span>
    <span class="o">-</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
<div class="footer" role="contentinfo" align="center">
    <a href="https://github.com/phyletica/ecoevolity">GitHub</a> &nbsp; | &nbsp; <a href="../license.html">License</a> &nbsp; | &nbsp; <a href="../citation.html">Citation</a> &nbsp; | &nbsp; <a href="https://groups.google.com/forum/#!forum/ecoevolity-users">Users Forum</a>
</div>

  </body>
</html>