
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>The Data &#8212; Ecoevolity 0.1.0 documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="The Configuration File" href="yaml-config.html" />
    <link rel="prev" title="Background" href="background.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../index.html">
    <img class="logo" src="../_static/site-logo.svg" alt="Logo"/>
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=phyletica&repo=ecoevolity&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ecoevolity/index.html">Ecoevolity</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Phycoeval</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="background.html">Background</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">The Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="yaml-config.html">The Configuration File</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../workshops/index.html">Workshops</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changes.html">Change History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citation.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../acknowledgments.html">Acknowledgments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zbib/references.html">Literature Cited</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html"></a><ul>
      <li>Previous: <a href="background.html" title="previous chapter">Background</a></li>
      <li>Next: <a href="yaml-config.html" title="next chapter">The Configuration File</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <p><a class="reference external" href="index.html"><img alt="Phycoeval" src="../_images/phycoeval-logo-long.svg" width="95%" /></a></p>
<div class="section" id="the-data">
<span id="phycodata"></span><h1>The Data<a class="headerlink" href="#the-data" title="Permalink to this headline">¶</a></h1>
<p>To use phycoeval, you will need <strong>biallelic</strong> genetic characters aligned across
the species for which you wish to infer a phylogeny.
Phycoeval assumes these characters are <strong>unlinked</strong>, but, depending on your
data, you might be better off violating this assumption of the model
(more about this <a class="reference internal" href="#phyco-unlinked"><span class="std std-ref">below</span></a>).
<strong>Biallelic</strong> means that each character has two possible states.
But, most genomic data are composed of nucleotides, which have four states.
This can be accommodated by
<a class="reference internal" href="#phyco-codingnucs"><span class="std std-ref">coding the nucleotides as biallelic</span></a>.</p>
<div class="section" id="coding-nucleotide-characters">
<span id="phyco-codingnucs"></span><h2>Coding nucleotide characters<a class="headerlink" href="#coding-nucleotide-characters" title="Permalink to this headline">¶</a></h2>
<p>Phycoeval will read your nucleotide characters, and recode them as biallelic.
If phycoeval finds a polyallelic character with more than two states across your
samples, it will report an error and stop running.
However, you can tell phycoeval to recode these polyallelic characters as either
being the first state (0) or a different state (1).
Your other option is to exclude these sites.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are analyzing nucleotide data as biallelic characters, we strongly
recommend that you do not try and estimate the frequencies of the two
states.
Instead, <a class="reference internal" href="yaml-config.html#phycofreq-1"><span class="std std-ref">fix the frequencies to be equal</span></a>.
This is because there are many ways to recode the 4 nucleotide states to two
states.
Thus, if you try to estimate frequencies of the two states, your results can be
sensitive to how you decided to code your nucleotides as binary.</p>
</div>
</div>
<div class="section" id="unlinked">
<span id="phyco-unlinked"></span><h2>Unlinked<a class="headerlink" href="#unlinked" title="Permalink to this headline">¶</a></h2>
<p><strong>Unlinked</strong> means that each character is assumed to have evolved along a gene
tree that is independent from the others (conditional on the population
history).
In other words, it assumes that the characters are far enough apart from one
another in the genome that they segregate independently during meiosis.</p>
<p>Many genomic data sets consist of many loci, each of which is a stretch of
linked nucleotides.
Examples include “RADseq” or sequence-capture data.
<a class="reference internal" href="#phyco-linkage"><span class="std std-ref">What should we do with such data</span></a>?</p>
<div class="section" id="what-to-do-with-linked-characters">
<span id="phyco-linkage"></span><h3>What to do with linked characters?<a class="headerlink" href="#what-to-do-with-linked-characters" title="Permalink to this headline">¶</a></h3>
<p>What if you have loci comprising sequences of linked nucleotides?
Based on simulations
<a class="bibtex reference internal" href="../zbib/references.html#oaks2018ecoevolity" id="id1">[14]</a><a class="bibtex reference internal" href="../zbib/references.html#oaks2018paic" id="id2">[18]</a><a class="bibtex reference internal" href="../zbib/references.html#oaks2019codemog" id="id3">[16]</a><a class="bibtex reference internal" href="../zbib/references.html#oaks2021phycoeval" id="id4">[20]</a>,
we recommend that you analyze all of your
characters (including the constant ones) and violate the assumption of unlinked
characters.
Ecoevolity and phycoeval perform better when you use all of the sites (including
linked and constant ones) compared to reducing the data to only one variable
character per locus.
So, your best bet is to include all of the sites from loci.</p>
</div>
</div>
<div class="section" id="formatting-your-data">
<span id="phyco-nexusfile"></span><h2>Formatting your data<a class="headerlink" href="#formatting-your-data" title="Permalink to this headline">¶</a></h2>
<p>Currently, phycoeval accepts two input formats for genetic characters:
Nexus and <a class="reference external" href="http://yaml.org/">YAML</a>.</p>
<div class="section" id="nexus-format">
<h3>Nexus format<a class="headerlink" href="#nexus-format" title="Permalink to this headline">¶</a></h3>
<div class="section" id="standard-haploid-data">
<h4>Standard haploid data<a class="headerlink" href="#standard-haploid-data" title="Permalink to this headline">¶</a></h4>
<p>You can represent your data in a “standard” 0/1 format.
Here’s an example of three species from which we’ve sampled 4 genomes each (2
diploid individuals):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN TAXA;
    DIMENSIONS NTAX=12;
    TAXLABELS
        lizard-953-a-speciesA
        lizard-953-b-speciesA
        lizard-954-a-speciesA
        lizard-954-b-speciesA
        lizard-152-a-speciesB
        lizard-152-b-speciesB
        lizard-154-a-speciesB
        lizard-154-b-speciesB
        lizard-331-a-speciesC
        lizard-331-b-speciesC
        lizard-338-a-speciesC
        lizard-338-b-speciesC
    ;
END;

BEGIN CHARACTERS;
    DIMENSIONS NCHAR=273658;
    FORMAT DATATYPE=STANDARD SYMBOLS=&quot;01&quot; MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  001010...
        lizard-953-b-speciesA  001010...
        lizard-954-a-speciesA  101010...
        lizard-954-b-speciesA  001011...
        lizard-152-a-speciesB  101110...
        lizard-152-b-speciesB  101110...
        lizard-154-a-speciesB  001011...
        lizard-154-b-speciesB  101010...
        lizard-331-a-speciesC  001010...
        lizard-331-b-speciesC  001010...
        lizard-338-a-speciesC  011010...
        lizard-338-b-speciesC  001011...
    ;
END;
</pre></div>
</div>
<p>Note, we don’t need separate TAXA and CHARACTER blocks like above.
Instead, we can specify a DATA block:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=12 NCHAR=273658;
    FORMAT DATATYPE=STANDARD SYMBOLS=&quot;01&quot; MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  001010...
        lizard-953-b-speciesA  001010...
        lizard-954-a-speciesA  101010...
        lizard-954-b-speciesA  001011...
        lizard-152-a-speciesB  101110...
        lizard-152-b-speciesB  101110...
        lizard-154-a-speciesB  001011...
        lizard-154-b-speciesB  101010...
        lizard-331-a-speciesC  001010...
        lizard-331-b-speciesC  001010...
        lizard-338-a-speciesC  011010...
        lizard-338-b-speciesC  001011...
    ;
END;
</pre></div>
</div>
<p>Both examples above would be equivalent for phycoeval, but the
<a class="reference external" href="http://ncl.sourceforge.net/">Nexus Class Library</a>
used by phycoeval will report a message about an implicit TAXA block if you use the
latter format.
Either way, in your <a class="reference internal" href="yaml-config.html#phycoconfigfile"><span class="std std-ref">phycoeval config file</span></a>,
you need to tell phycoeval that the
<a class="reference internal" href="yaml-config.html#phycogenotypesarediploid"><span class="std std-ref">states, or genotypes, are haploid</span></a>
by declaring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">false</span>
</pre></div>
</div>
</div>
<div class="section" id="standard-diploid-data">
<h4>Standard diploid data<a class="headerlink" href="#standard-diploid-data" title="Permalink to this headline">¶</a></h4>
<p>Above, each cell in our matrix represented which state was present
for the character in a particular haploid genome.
We can also represent the same data where each cell represents
the genotype of a diploid individual:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=6 NCHAR=273658;
    FORMAT DATATYPE=STANDARD SYMBOLS=&quot;012&quot; MISSING=? GAP=-;
    MATRIX
        lizard-953-speciesA  002020...
        lizard-954-speciesA  102021...
        lizard-152-speciesB  202220...
        lizard-154-speciesB  102021...
        lizard-331-speciesC  002020...
        lizard-338-speciesC  012021...
    ;
END;
</pre></div>
</div>
<p>Now, “0” represents that the individual has two copies with the 0 state, “2”
represents two copies of the 1 state, and “1” represents a heterozygote.
Again, in your <a class="reference internal" href="yaml-config.html#phycoconfigfile"><span class="std std-ref">phycoeval config file</span></a>,
you need to tell phycoeval that the
<a class="reference internal" href="yaml-config.html#phycogenotypesarediploid"><span class="std std-ref">states, or genotypes, are diploid</span></a>
by declaring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">true</span>
</pre></div>
</div>
</div>
<div class="section" id="nucleotide-data">
<h4>Nucleotide data<a class="headerlink" href="#nucleotide-data" title="Permalink to this headline">¶</a></h4>
<p>If you have nucleotide data, the easiest thing is provide the nucleotide
characters to phycoeval as is, and let it recode them as biallelic.
Here’s an example where we are providing nucleotides as haploid (each cell is a
haploid genotype):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=12 NCHAR=273658;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  ACGTAG...
        lizard-953-b-speciesA  ACGTAG...
        lizard-954-a-speciesA  GCGTAG...
        lizard-954-b-speciesA  ACGTAA...
        lizard-152-a-speciesB  GCGCAG...
        lizard-152-b-speciesB  GCGCAG...
        lizard-154-a-speciesB  ACGTAA...
        lizard-154-b-speciesB  GCGTAG...
        lizard-331-a-speciesC  ACGTAG...
        lizard-331-b-speciesC  ACGTAG...
        lizard-338-a-speciesC  ATGTAG...
        lizard-338-b-speciesC  ACGTAA...
    ;
END;
</pre></div>
</div>
<p>This is sometimes referred to as “phased” data.
Again, if we are providing a matrix where each cell represents a haploid
genotype, we need to tell phycoeval this is so via the
<a class="reference internal" href="yaml-config.html#phycoconfigfile"><span class="std std-ref">config file</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">false</span>
</pre></div>
</div>
<p>We can also represent the same data as “unphased”, where each cell represents a
diploid genotype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=6 NCHAR=273658;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  ACGTAG...
        lizard-954-a-speciesA  RCGTAR...
        lizard-152-a-speciesB  GCGCAG...
        lizard-154-a-speciesB  RCGTAR...
        lizard-331-a-speciesC  ACGTAG...
        lizard-338-a-speciesC  AYGTAR...
    ;
END;
</pre></div>
</div>
<p>We need to indicate this in the
<a class="reference internal" href="../ecoevolity/yaml-config.html#configfile"><span class="std std-ref">config file</span></a>::
accordingly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genotypes_are_diploid</span><span class="p">:</span> <span class="n">true</span>
</pre></div>
</div>
</div>
<div class="section" id="population-labels">
<h4>Population labels<a class="headerlink" href="#population-labels" title="Permalink to this headline">¶</a></h4>
<p>In our nexus character matrix, we need to indicate which species (or
population) each row corresponds to.
We can do this with either using a prefix or suffix in the row (or taxon)
labels.
For example, in this nexus data file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=6 NCHAR=273658;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  ACGTAG...
        lizard-954-a-speciesA  RCGTAR...
        lizard-152-a-speciesB  GCGCAG...
        lizard-154-a-speciesB  RCGTAR...
        lizard-331-a-speciesC  ACGTAG...
        lizard-338-a-speciesC  AYGTAR...
    ;
END;
</pre></div>
</div>
<p>we are using the suffixes to indicate that the first two samples came
from a species we are calling <code class="docutils literal notranslate"><span class="pre">speciesA</span></code>,
the next to samples came from a species called <code class="docutils literal notranslate"><span class="pre">speciesB</span></code>,
and
the last two samples came from a species we are calling
<code class="docutils literal notranslate"><span class="pre">speciesC</span></code>.
In our phycoeval <a class="reference internal" href="yaml-config.html#phycoconfigfile"><span class="std std-ref">config file</span></a>
we have to indicate this with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">population_name_delimiter</span><span class="p">:</span> <span class="s2">&quot;-&quot;</span>
<span class="n">population_name_is_prefix</span><span class="p">:</span> <span class="n">false</span>
</pre></div>
</div>
<p>This tells phycoeval to look for the last bit of each row label that is
separated by a “-” to figure out the population label.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you like to use underscores as a population label
delimiter, just watch out for a
<a class="reference internal" href="yaml-config.html#phycounderscoregotcha"><span class="std std-ref">gotcha related to how the nexus format treats underscores</span></a></p>
</div>
</div>
</div>
<div class="section" id="yaml-format">
<h3>YAML format<a class="headerlink" href="#yaml-format" title="Permalink to this headline">¶</a></h3>
<p>The YAML-formatted data format is a lot more efficient (much smaller file
sizes).
Instead of a full alignment,
it only contains a list of allele count patterns, followed by a list of the
weight of each pattern (i.e., how many times the allele pattern occurs in the
alignment).
For a very small example let’s convert the following nexus alignment
and convert it to the <a class="reference external" href="http://yaml.org/">YAML</a> format that ecoevolity accepts as input:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS

BEGIN DATA;
    DIMENSIONS NTAX=12 NCHAR=6;
    FORMAT DATATYPE=DNA MISSING=? GAP=-;
    MATRIX
        lizard-953-a-speciesA  ACGTAG
        lizard-953-b-speciesA  ACGTAG
        lizard-954-a-speciesA  GCGTAG
        lizard-954-b-speciesA  ACGTAA
        lizard-152-a-speciesB  GCGCAG
        lizard-152-b-speciesB  GCGCAG
        lizard-154-a-speciesB  ACGTAA
        lizard-154-b-speciesB  GCGTAG
        lizard-331-a-speciesC  ACGTAG
        lizard-331-b-speciesC  ACGTAG
        lizard-338-a-speciesC  ATGTAG
        lizard-338-b-speciesC  ACGTAA
    ;
END;
</pre></div>
</div>
<p>To convert these data to <a class="reference external" href="http://yaml.org/">YAML</a> format, we will assume the first nucleotide
(from the top) in each column is state “0”, and the second nucleotide (if any)
is state “1”.
Doing so gives us the following allele-count patterns in <a class="reference external" href="http://yaml.org/">YAML</a> format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">---</span>
<span class="n">markers_are_dominant</span><span class="p">:</span> <span class="n">false</span>
<span class="n">population_labels</span><span class="p">:</span>
    <span class="o">-</span> <span class="n">speciesA</span>
    <span class="o">-</span> <span class="n">speciesB</span>
    <span class="o">-</span> <span class="n">speciesC</span>
<span class="n">allele_count_patterns</span><span class="p">:</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
    <span class="o">-</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">]]</span>
<span class="n">pattern_weights</span><span class="p">:</span>
    <span class="o">-</span> <span class="mi">1</span>
    <span class="o">-</span> <span class="mi">1</span>
    <span class="o">-</span> <span class="mi">2</span>
    <span class="o">-</span> <span class="mi">1</span>
    <span class="o">-</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
<div class="footer" role="contentinfo" align="center">
    <a href="https://github.com/phyletica/ecoevolity">GitHub</a> &nbsp; | &nbsp; <a href="../license.html">License</a> &nbsp; | &nbsp; <a href="../citation.html">Citation</a> &nbsp; | &nbsp; <a href="https://groups.google.com/forum/#!forum/ecoevolity-users">Users Forum</a>
</div>

  </body>
</html>